### PR TITLE
Pre increment

### PR DESCRIPTION
**Pull request by shrit**
These two commits are a cherry-pick from #2415 

Pre-increment is faster than post-increment
I know that gcc generated the same code, but it is easier to keep a good practice

### CODE REVIEW

**Comment by jeffin143** [comment]
did you use some kind of grep or awk to replace all the instances in files ?

**Comment by shrit** [comment]
Yes, I use sed for this kind of stuff

File: src/mlpack/core/math/columns_to_blocks.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ void ColumnsToBlocks::Transform(const arma::mat& maximalInputs,
       const size_t maxCol = j * colOffset + blockWidth;
 
       output.submat(minRow, minCol, maxRow, maxCol) =
-          arma::reshape(maximalInputs.col(k++), blockHeight, blockWidth);
+          arma::reshape(maximalInputs.col(++k), blockHeight, blockWidth);
```
Comment (birm) [inline src/mlpack/core/math/columns_to_blocks.cpp:80]: I think that this is the cause of https://dev.azure.com/mlpack/mlpack/_build/results?buildId=2671&view=logs&j=24d3abe3-ef0b-5deb-3aab-64d839de2c3c&t=b3d23cc5-b695-5043-0f03-2084bf2ff0b5&l=155

Comment (shrit) [reply src/mlpack/core/math/columns_to_blocks.cpp:80]: Right, in this case, we need to use post-increment,

File: src/mlpack/methods/ann/callbacks/early_stop_at_min_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ * decreasing.
+ */
+class EarlyStopAtMinLoss
+{
+ public:
+  /**
+   * Set up the early stop at min loss class, which keeps track of the minimum
+   * loss and stops the optimization process if the loss stops decreasing.
+   *
+   * @param patienceIn The number of epochs to wait after the minimum loss has
```
Comment (birm) [inline src/mlpack/methods/ann/callbacks/early_stop_at_min_loss.hpp:30]: It looks like you need to document output too.

Comment (shrit) [reply src/mlpack/methods/ann/callbacks/early_stop_at_min_loss.hpp:30]: Actually, this file is added by error, I have to remove it.

File: src/mlpack/core/tree/cosine_tree/cosine_tree.cpp [hunk 3]
```diff
@@ -451,7 +451,7 @@ double CosineTree::MonteCarloError(CosineTree* node,
     // If two additional vectors are passed, take their projections.
     if (addBasisVector1 && addBasisVector2)
     {
-      projection(k++) = arma::dot(dataset.col(sampledIndices[i]),
+      projection(++k) = arma::dot(dataset.col(sampledIndices[i]),
```
Comment (rcurtin) [inline src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:454]: This one probably won't work as-is.

File: src/mlpack/methods/ann/ann_low_resource_main.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
+/**
```
Comment (rcurtin) [inline src/mlpack/methods/ann/ann_low_resource_main.cpp:1]: Oops, I think we don't need to commit this file. :+1:

File: src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
+/**
```
Comment (rcurtin) [inline src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp:1]: This file looks like it was added by accident too. :)

File: src/mlpack/tests/main_tests/hmm_train_test.cpp [hunk 6]
```diff
@@ -64,7 +64,7 @@ inline void CheckMatricesDiffer(arma::mat& a, arma::mat& b, double tolerance)
   bool valsEqual = true;
   if (dimsEqual)
   {
-    for (size_t i=0; i<a.n_elem; i++)
+    for (size_t i=0; i<a.n_elem; ++i)
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/hmm_train_test.cpp:67]:
```suggestion
    for (size_t i = 0; i < a.n_elem; ++i)
```
Just a style fix I saw while looking through. :)

File: src/mlpack/methods/ann/ann_low_resource_main.cpp (hunk 4 above)
Comment (shrit) [reply src/mlpack/methods/ann/ann_low_resource_main.cpp:1]: Sorry for this,  it has accidentally escaped from mlpack_rpi

File: src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp (hunk 5 above)
Comment (shrit) [reply src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp:1]: Sorry for this one too

**Comment by shrit** [comment]
I think this one can be merged since it is compiling

**Comment by rcurtin** [comment]
@shrit looks good to me, but since you touched so many files it set off some static code analysis issues.  Would you be willing to fix them "while you're at it"?

I took a look through them:

1. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.1225452494/source.c7c1fd4e-e190-430a-ba06-93d5b5a2514d/#162

Looks like the static code analysis checker is compiling without `ARMA_USE_HDF5` defined.  That's ok... we can change the code at line 161 to

```
#ifdef ARMA_USE_HDF5
    // We can't save with streams for HDF5.
    const bool success = (saveType == arma::hdf5_binary) ?
        tmp.quiet_save(filename, saveType) :
        tmp.quiet_save(stream, saveType);
#else
    const bool success = tmp.quiet_save(stream, saveType);
#endif
```

2. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.1225452494/source.d44cfe76-752f-4d37-882b-f0cd394838c1/#179

Same exact change here, but at line 179:

```
#ifdef ARMA_USE_HDF5
    // We can't save with streams for HDF5.
    const bool success = (saveType == arma::hdf5_binary) ?
        matrix.quiet_save(filename, saveType) :
        matrix.quiet_save(stream, saveType);
#else
    const bool success = matrix.quiet_save(stream, saveType);
#endif
```

3. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.-665440279/source.3df42331-8598-42f4-9f1c-58a4354fae91/#20

Looks like `logDetCov` is not initialized in the constructor.  You could just modify the constructor for `GaussianDistribution`:

```
GaussianDistribution::GaussianDistribution(const arma::vec& mean,
                                           const arma::mat& covariance)
  : mean(mean), logDetCov(0.0)
{
  Covariance(covariance);
}
```

Realistically, `logDetCov` is set in `Covariance()`, so this is a spurious warning but it's easy to fix.

4. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.1225945026/fileName.1122149002/type.100709/

Ooh, this one is actually a memory leak!  It's easy to fix; at line 90 in `cellbound_impl.hpp` we can replace the code with this:

```
  if (dim != other.Dim())
  {
    // Reallocation is necessary.
    delete[] bounds;

    dim = other.Dim();
    bounds = new math::RangeType<ElemType>[dim];
  }
```

5. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.1225945026/fileName.1122149002/type.3641990/

Ah, we just need to check for equality in the copy constructor.  We can just add this to the top of the function (cellbound_impl.hpp:89):

```
  if (this == &other)
    return *this;
```

6. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.1225945026/fileName.100195514/source.96c7d9d5-6868-4023-8012-de435c422cfc/#67

Same issue here; we can put this at the top of the function at hrectbound_impl.hpp:67:

```
  if (this == &other)
    return this;
```

7. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.322822747/fileName.-1383868599/

This looks like another memory leak, but this one is a bit tricky because we may not always allocate space for `localHilbertValues`.  So in this case we should do something a little more careful at line 438:

```
  if (ownsLocalHilbertValues)
    delete localHilbertValues;

  localHilbertValues = const_cast<arma::Mat<HilbertElemType>* >
      (val.LocalHilbertValues());
  ownsLocalHilbertValues = false;
  numValues = val.NumValues();
```

8. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.322822747/fileName.677819056/source.9674d3dc-6784-4c84-9fd9-8b7a2e4263bc/#79

Hehe, this `assert()` is unnecessary then and we can just remove it.

9. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.322822747/fileName.677819056/source.8db4ad3c-9fa7-40e6-8200-2bb81ee5a77e/#144

Same here.

10. http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5448/cppcheck/folder.322822747/fileName.677819056/source.b7f0ee7e-a6eb-462c-925d-e99e56ac4c6a/#206

And one more, same thing.

If you're willing to fix those, it would be awesome; otherwise I can find a minute or two to make these changes and push them to the branch. :+1:

**Comment by shrit** [comment]
Yes, of course, I can do that today

**Comment by shrit** [comment]
@rcurtin Those are fixed, when the check are done we can merge with out any issue :+1:

**Comment by zoq** [comment]
> @shrit looks good to me, but since you touched so many files it set off some static code analysis issues. Would you be willing to fix them "while you're at it"?

Wow, this is a nice list of patches.

**Comment by rcurtin** [comment]
@shrit looks like there is just one more:
http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/5482/cppcheck/source.2b530391-1209-4f99-8069-1f4507ca7db1/#438

It seems pretty simple, just this is needed in `discrete_hilbert_value_impl.hpp` at line 439:

```
if (this == &val)
  return *this;
```

**Comment by rcurtin** [comment]
Awesome, everything is working.  Thanks @shrit for making these changes and fixing all the static code analysis errors too! :+1:

**Pull request merged**